File: 054-spiral-matrix/spiral-matrix.py
# -*- coding:utf-8 -*-


# Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order.
#
#
#
# For example,
# Given the following matrix:
#
#
# [
#  [ 1, 2, 3 ],
#  [ 4, 5, 6 ],
#  [ 7, 8, 9 ]
# ]
#
#
# You should return [1,2,3,6,9,8,7,4,5].


class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        m = len(matrix)
        if m == 0: return []
        n = len(matrix[0])
        res = [0 for i in range(n*m)]
        next_dir = {0: 1, 1:2, 2:3, 3:0}  # 0: right; 1: down; 2: left; 3: up
        d = {0: (0, 1), 1: (1, 0), 2: (0, -1), 3: (-1, 0)}  # (y, x)
        curx = 0
        cury = 0
        curdir = 0
        for i in range(1, m*n + 1):
            res[i - 1] = matrix[cury][curx]
            matrix[cury][curx] = 0
            attempt_x = curx + d[curdir][1]
            attempt_y = cury + d[curdir][0]
            if attempt_x >= n or attempt_y >= m or attempt_x < 0 or attempt_y < 0 or matrix[attempt_y][attempt_x] == 0:
                curdir = next_dir[curdir]
            curx = curx + d[curdir][1]
            cury = cury + d[curdir][0]
        return res
        

File: 054-spiral-matrix/spiral-matrix.cpp
// Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order.
//
// Example 1:
//
//
// Input:
// [
//  [ 1, 2, 3 ],
//  [ 4, 5, 6 ],
//  [ 7, 8, 9 ]
// ]
// Output: [1,2,3,6,9,8,7,4,5]
//
//
// Example 2:
//
// Input:
// [
//   [1, 2, 3, 4],
//   [5, 6, 7, 8],
//   [9,10,11,12]
// ]
// Output: [1,2,3,4,8,12,11,10,9,5,6,7]
//


class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;
        int height = matrix.size();
        if (height == 0) return res;
        int width = matrix[0].size();
        int status = 0; // 0: go_right; 1: go_down; 2: go_left; 3: go_up
        //int limits[4] = {width - 1, height - 1, 0, 0};
        int left = 0, right = width - 1;
        int up = 0, down = height - 1;
        //int next_dir[4] = {1, 2, 3, 0};
        int dirs[4][2] = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}}; // x, y
        int x = 0, y = 0;
        while (true) {
            res.push_back(matrix[y][x]);
            int nx = x + dirs[status][0];
            int ny = y + dirs[status][1];
            if (status == 0 && nx > right) {
                nx = x; ny = y + 1;
                if (ny > down) break;
                status = 1;
                up++;
            }
            else if (status == 1 && ny > down) {
                nx = x - 1; ny = y;
                if (nx < left) break;
                status = 2;
                right--;
            }
            else if (status == 2 && nx < left) {
                nx = x; ny = y - 1;
                if (ny < up) break;
                status = 3;
                down--;
            }
            else if (status == 3 && ny < up) {
                nx = x + 1; ny = y;
                if (nx > right) break;
                status = 0;
                left++;
            }
            x = nx; y = ny;
        }
        return res;
    }
};
